### PR TITLE
Issue 15, implement Orders

### PR DESCRIPTION
**Pull request by cbj4074**

### CODE REVIEW

File: src/Repositories/Eloquent/OrderRepository.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        return Order::class;
+    }
+
+    public function model(): Model
+    {
+        return $this->order;
+    }
+
+    public function delete(int $id): bool
```
Comment (poppabear8883) [inline src/Repositories/Eloquent/OrderRepository.php]: This method could probably be omitted if you merge in the changes from PR #16

Comment (cbj4074) [reply src/Repositories/Eloquent/OrderRepository.php]: Got it; I removed the method and indeed all tests still pass. Thanks!

File: database/migrations/2020_01_13_000001_create_order_products_table.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('order_product', function (Blueprint $table) {
+            $table->bigIncrements('id');
```
Comment (poppabear8883) [inline database/migrations/2020_01_13_000001_create_order_products_table.php]: this doesn't need a primary key if its a pivot table ?

File: database/migrations/2020_01_13_000001_create_order_products_table.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            $table->unsignedBigInteger('order_id')->index();
+            $table->foreign('order_id')->references('id')->on('orders');
+            $table->unsignedBigInteger('product_id')->unsigned()->index();
+            $table->foreign('product_id')->references('id')->on('products');
```
Comment (poppabear8883) [inline database/migrations/2020_01_13_000001_create_order_products_table.php:21]: I personally prefer this style just due to readability:

```php
$table->unsignedBigInteger('order_id');
$table->unsignedBigInteger('product_id');

$table->foreign('order_id')->references('id')->on('orders');
$table->foreign('product_id')->references('id')->on('products');
```

File: database/migrations/2020_01_13_000001_create_order_products_table.php [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('order_product', function (Blueprint $table) {
+            $table->bigIncrements('id');
+            $table->unsignedBigInteger('order_id')->index();
+            $table->foreign('order_id')->references('id')->on('orders');
+            $table->unsignedBigInteger('product_id')->unsigned()->index();
```
Comment (poppabear8883) [inline database/migrations/2020_01_13_000001_create_order_products_table.php]: You don't need the `->unsigned()->index();` 

`unsignedBigInteger` already applies `unsigned` and `index`

File: database/migrations/2020_01_13_000001_create_order_products_table.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('order_product', function (Blueprint $table) {
+            $table->bigIncrements('id');
+            $table->unsignedBigInteger('order_id')->index();
```
Comment (poppabear8883) [inline database/migrations/2020_01_13_000001_create_order_products_table.php:17]: You don't need the `->index()`

`unsigned` already applies a `index`

File: grumphp.yml [hunk 6]
```diff
@@ -52,12 +52,11 @@ parameters:
         - "die("
         - "var_dump("
         - "exit;"
-        - "exit("
```
Comment (poppabear8883) [inline grumphp.yml:55]: This is simply commented out in PR #26 

We can resolve this either way.

File: src/Repositories/Eloquent/OrderRepository.php [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+    /**
+     * OrderRepository constructor.
+     * @param Order $order
+     * @param DatabaseManager $db
+     */
+    public function __construct(Order $order, DatabaseManager $db)
+    {
+        $this->order = $order;
+        $this->db = $db;
```
Comment (poppabear8883) [inline src/Repositories/Eloquent/OrderRepository.php]: `DatabaseManager` is unused. You should consider removing it until it is required to pass a test.

**Review by poppabear8883 — CHANGES_REQUESTED** [review]
Some comments regarding some suggested changes. Will review after those have been resolved.

File: database/migrations/2020_01_13_000001_create_order_products_table.php (hunk 2 above)
Comment (cbj4074) [reply database/migrations/2020_01_13_000001_create_order_products_table.php]: Good catch. I copied this migration from https://github.com/Laracommerce/laracom/blob/5cdddc4bc5aa15b4791799c224892b09d8c6a5d4/database/migrations/2017_06_13_091740_create_order_products_table.php .

File: database/migrations/2020_01_13_000001_create_order_products_table.php (hunk 3 above)
Comment (cbj4074) [reply database/migrations/2020_01_13_000001_create_order_products_table.php:21]: I agree that it's easier to read and will make that change.

File: database/migrations/2020_01_13_000001_create_order_products_table.php (hunk 4 above)
Comment (cbj4074) [reply database/migrations/2020_01_13_000001_create_order_products_table.php]: Are you sure that it applies `index`, too?

Comment (poppabear8883) [reply database/migrations/2020_01_13_000001_create_order_products_table.php]: you know .. im not sure .. i thought i was sure but after digging into the source i can't see where it does.

File: database/migrations/2020_01_13_000001_create_order_products_table.php (hunk 5 above)
Comment (poppabear8883) [reply database/migrations/2020_01_13_000001_create_order_products_table.php:17]: Apparently i am wrong here, you can ignore this

**Comment by cbj4074** [comment]
I'm removing WIP status because I think it makes sense to add the other relations, such as Address, Courier, and Order Status in a subsequent PR. I just don't want this branch to diverge so far (especially in consideration of the other refactoring that's happening) that it becomes a nightmare to merge back in.

File: database/migrations/2020_01_13_000000_create_orders_table.php [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+     */
+    public function up()
+    {
+        Schema::create('orders', function (Blueprint $table) {
+            $table->bigIncrements('id');
+            $table->string('reference')->unique()->index();
+            //$table->unsignedInteger('courier_id')->index();
+            $table->unsignedBigInteger('customer_id')->index()->nullable();
+
+            if (Schema::hasTable(config('user_table'))) {
```
Comment (poppabear8883) [inline database/migrations/2020_01_13_000000_create_orders_table.php]: This is likely going to fail due to the lack of a configuration file.

Comment (cbj4074) [reply database/migrations/2020_01_13_000000_create_orders_table.php]: Oops, nice catch. That should be `config('velkart.user_table', 'users')`; I'll fix it now.

Comment (poppabear8883) [reply database/migrations/2020_01_13_000000_create_orders_table.php]: also, the syntax is wrong for the config path

**Review by poppabear8883 — APPROVED** [review]

**Pull request merged**